cedar-policy-core/src/tpe/evaluator.rs

```diff
@@ -0,0 +1,1667 @@
+/*
+ * Copyright Cedar Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! This module contains the type-aware partial evaluator.
+
+use std::{collections::BTreeMap, sync::Arc};
+
+use crate::validator::types::Type;
+use crate::{
+    ast::{self, BinaryOp, EntityUID, Expr, ExprKind, PartialValue, Set, Value, ValueKind, Var},
+    extensions::Extensions,
+};
+
+use crate::{
+    tpe::entities::PartialEntities,
+    tpe::request::PartialRequest,
+    tpe::residual::{Residual, ResidualKind},
+};
+
+/// The partial evaluator
+#[derive(Debug)]
+pub struct Evaluator<'e> {
+    pub(crate) request: PartialRequest,
+    pub(crate) entities: &'e PartialEntities,
+    pub(crate) extensions: &'e Extensions<'e>,
```

Comment:
tiny nit: `request` could also be a `&'e` reference. I notice that at least one caller of `Evaluator` ends up having to (needlessly?) clone their `PartialRequest` because of this.